src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,553 @@
+            ["/usr/bin/virt-sparsify", "--compress", str(image), str(IMAGE_OUTPUT_PATH)],
+            check=True,
+            timeout=60 * 10,
+        )
+    except subprocess.CalledProcessError as exc:
+        raise ImageCompressError from exc
+
+
+@dataclasses.dataclass
+class BuildImageConfig:
+    """Configuration for building the image.
+
+    Attributes:
+        arch: The CPU architecture to build the image for.
+        base_image: The ubuntu image to use as build base.
+    """
+
+    arch: Arch
+    base_image: BaseImage
+
+
+def build_image(config: BuildImageConfig) -> None:
+    """Build and save the image locally.
+
+    Args:
+        config: The configuration values to build the image with.
+
+    Raises:
+        BuildImageError: If there was an error building the image.
+    """
+    logger.info("Clean build state.")
+    with redirect_stdout(sys.stderr):
+        try:
+            _clean_build_state()
+            logger.info("Downloading cloud image.")
+            cloud_image_path = _download_cloud_image(
+                arch=config.arch, base_image=config.base_image
+            )
+            logger.info("Resizing cloud image.")
+            _resize_cloud_img(cloud_image_path=cloud_image_path)
+            logger.info("Mounting network block device.")
+            _mount_image_to_network_block_device(cloud_image_path=cloud_image_path)
+            logger.info("Replacing resolv.conf.")
+            _replace_mounted_resolv_conf()
+            logger.info("Resizing partitions.")
+            _resize_mount_partitions()
+            logger.info("Building YQ from source.")
+            _install_yq()
+        except ImageBuilderBaseError as exc:
+            raise BuildImageError from exc
+
+        try:
+            logger.info("Setting up chroot environment.")
+            with ChrootContextManager(IMAGE_MOUNT_DIR):
+                # operator_libs_linux apt package uses dpkg -l and that does not work well with
+                # chroot env, hence use subprocess run.
+                subprocess.run(
+                    ["/usr/bin/apt-get", "update", "-y"],
+                    check=True,
+                    timeout=60 * 10,
+                    env={"DEBIAN_FRONTEND": "noninteractive"},
+                )  # nosec: B603
+                subprocess.run(  # nosec: B603
+                    [
+                        "/usr/bin/apt-get",
+                        "install",
+                        "-y",
+                        "--no-install-recommends",
+                        *IMAGE_DEFAULT_APT_PACKAGES,
+                    ],
+                    check=True,
+                    timeout=60 * 20,
+                    env={"DEBIAN_FRONTEND": "noninteractive"},
+                )
+                _create_python_symlinks()
+                _disable_unattended_upgrades()
+                _configure_system_users()
+                _install_external_packages()
+        except (ImageBuilderBaseError, ChrootBaseError) as exc:
```

Comment:
finally:
  clean build state

Reply:
Decided not to clean build state after runs, only at the beginning of the run.
When exceptions are re-raised, the clean build state before raising error every time seems unnecessary.